dartagnan/src/main/java/com/dat3m/dartagnan/witness/graphviz/ExecutionGraphVisualizer.java

```diff
@@ -188,18 +186,36 @@ private ExecutionGraphVisualizer addThreadPo(Thread thread, ExecutionModel model
         return this;
     }
 
+    private String getAddressString(BigInteger address) {
+        MemoryObject obj = null;
+        BigInteger objAddress = null;
+        for (Map.Entry<MemoryObject, BigInteger> entry : objToAddrMap.entrySet()) {
+            final BigInteger nextObjAddr = entry.getValue();
+            if (nextObjAddr.compareTo(address) > 0) {
+                break;
+            }
+            obj = entry.getKey();
+            objAddress = nextObjAddr;
+        }
+
+        if (obj == null) {
+            return address + " [OOB]";
+        } else if (address.equals(objAddress)) {
+            return obj.toString();
+        } else {
+            final boolean isOOB = address.compareTo(objAddress.add(BigInteger.valueOf(obj.size()))) >= 0;
```

Comment:
I don't get this check .... seems to detect some "non-trivial" OOB

Reply:
This check captures accesses that seem to belong to object `A` but actually go beyond `A`.
There are two cases where this happens:
(1) You have two adjacent memory objects `A` and `B` that have a gap inbetween them (say, due to alignment requirements), and the accessed address is precisely in this gap. The code will first think that the access goes to `A` but this additonal check figures out that the access really goes into the "gap".
(2) The access goes far beyond the last memory object `Z`, i.e., `Z + sizeof(Z) <= addr`. Again, the previous code will think it is an access to `Z` and the additional check sees that it actually goes beyond `Z` and thus is OOB.